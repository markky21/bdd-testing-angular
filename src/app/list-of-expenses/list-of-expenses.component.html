<mat-card class="Wrapper">
  <mat-card-header>
    <mat-card-title>List of expenses</mat-card-title>
  </mat-card-header>
  <mat-card-content>
    <app-month-selector></app-month-selector>
    <app-expense-table></app-expense-table>
    <app-expense-form></app-expense-form>
  </mat-card-content>
</mat-card>
