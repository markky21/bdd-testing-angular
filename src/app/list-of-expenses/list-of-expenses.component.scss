.Form {
  margin: 32px;
}
