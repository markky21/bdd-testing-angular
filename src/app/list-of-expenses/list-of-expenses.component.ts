import { Component } from '@angular/core';

@Component({
  selector: 'app-list-of-expenses',
  templateUrl: './list-of-expenses.component.html',
  styleUrls: ['./list-of-expenses.component.scss'],
})
export class ListOfExpensesComponent {}
