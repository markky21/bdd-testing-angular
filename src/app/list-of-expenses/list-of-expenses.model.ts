export enum ExpenseCategory {
  HOME = 'HOME',
  CAR = 'CAR',
}

export interface Expense {
  id?: number;
  date: string;
  amount: number;
  category: ExpenseCategory;
}
