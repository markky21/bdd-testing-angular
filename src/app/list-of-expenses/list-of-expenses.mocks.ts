import * as moment from 'moment';
import { Expense, ExpenseCategory } from './list-of-expenses.model';

export const mockExpenseList: Expense[] = [
  {
    id: 1,
    date: moment().add(-5, 'days').toISOString(),
    amount: 15.5,
    category: ExpenseCategory.CAR,
  },
  {
    id: 2,
    date: moment().add(-9, 'days').toISOString(),
    amount: 163.5,
    category: ExpenseCategory.HOME,
  },
  {
    id: 3,
    date: moment().add(-10, 'days').toISOString(),
    amount: 5.5,
    category: ExpenseCategory.HOME,
  },
];

export const getMockExpenseList: () => Expense[] = () => [
  {
    id: 1,
    date: moment().add(-5, 'days').toISOString(),
    amount: 15.5,
    category: ExpenseCategory.CAR,
  },
  {
    id: 2,
    date: moment().add(-9, 'days').toISOString(),
    amount: 163.5,
    category: ExpenseCategory.HOME,
  },
  {
    id: 3,
    date: moment().add(-10, 'days').toISOString(),
    amount: 5.5,
    category: ExpenseCategory.HOME,
  },
];
