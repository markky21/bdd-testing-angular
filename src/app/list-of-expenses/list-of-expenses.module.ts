import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ListOfExpensesComponent } from './list-of-expenses.component';
import { MatCardModule } from '@angular/material/card';
import { MatButtonModule } from '@angular/material/button';
import { ExpenseTableComponent } from './expense-table/expense-table.component';
import { MonthSelectorComponent } from './month-selector/month-selector.component';
import { MatFormFieldModule } from '@angular/material/form-field';
import { MatInputModule } from '@angular/material/input';
import { MatDatepickerModule } from '@angular/material/datepicker';
import { MatNativeDateModule } from '@angular/material/core';
import { ReactiveFormsModule } from '@angular/forms';
import { MatTableModule } from '@angular/material/table';
import { ExpenseFormComponent } from './expense-form/expense-form.component';
import {MatSelectModule} from '@angular/material/select';

@NgModule({
  declarations: [ListOfExpensesComponent, ExpenseTableComponent, MonthSelectorComponent, ExpenseFormComponent],
  imports: [
    CommonModule,
    MatCardModule,
    MatButtonModule,
    MatFormFieldModule,
    MatInputModule,
    MatDatepickerModule,
    MatNativeDateModule,
    ReactiveFormsModule,
    MatTableModule,
    MatSelectModule,
  ],
  exports: [ListOfExpensesComponent],
})
export class ListOfExpensesModule {}
