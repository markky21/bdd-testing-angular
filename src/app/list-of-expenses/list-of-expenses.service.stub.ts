import { ListOfExpensesService } from './list-of-expenses.service';
import { Expense } from './list-of-expenses.model';
import { asyncScheduler, BehaviorSubject, Observable, of, scheduled } from 'rxjs';
import { mockExpenseList } from './list-of-expenses.mocks';
import moment from 'moment';

export class ListOfExpensesServiceStub implements Pick<ListOfExpensesService, keyof ListOfExpensesService> {
  selectedMonth$ = new BehaviorSubject<moment.Moment | null>(moment());
  dbAdd$(expense: Expense): Observable<Expense[]> {
    return scheduled(of(mockExpenseList), asyncScheduler);
  }
  dbGetAllFilteredBySelectedMonth$(): Observable<Expense[]> {
    return scheduled(of(mockExpenseList), asyncScheduler);
  }
  getExpensesTotalAmount$(): Observable<number> {
    return scheduled(of(123), asyncScheduler);
  }
}
