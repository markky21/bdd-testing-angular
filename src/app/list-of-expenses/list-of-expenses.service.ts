import { Injectable } from '@angular/core';
import { BehaviorSubject, combineLatest, concat, Observable, Subject } from 'rxjs';
import moment from 'moment';
import { NgxIndexedDBService, ObjectStoreMeta } from 'ngx-indexed-db';
import { Expense } from './list-of-expenses.model';
import { filter, map, mergeMap, take, tap } from 'rxjs/operators';

const storeSchema: ObjectStoreMeta = {
  store: 'expenses',
  storeConfig: { keyPath: 'id', autoIncrement: true },
  storeSchema: [
    { name: 'date', keypath: 'expenseDate', options: { unique: false } },
    { name: 'amount', keypath: 'amount', options: { unique: false } },
    { name: 'category', keypath: 'amount', options: { unique: false } },
  ],
};

@Injectable({
  providedIn: 'root',
})
export class ListOfExpensesService {
  private abAll$ = new Subject<Expense[]>();
  selectedMonth$ = new BehaviorSubject<moment.Moment | null>(null);

  constructor(private ngxIndexedDBService: NgxIndexedDBService<Expense>) {
    ngxIndexedDBService.createObjectStore(storeSchema);
  }

  dbAdd$(expense: Expense): Observable<Expense[]> {
    return this.ngxIndexedDBService.add(storeSchema.store, expense).pipe(
      mergeMap(() => this.ngxIndexedDBService.getAll(storeSchema.store)),
      tap((value) => this.abAll$.next(value)),
      take(1)
    );
  }

  private dbGetAll$(): Observable<Expense[]> {
    return concat(this.ngxIndexedDBService.getAll(storeSchema.store), this.abAll$);
  }

  dbGetAllFilteredBySelectedMonth$(): Observable<Expense[]> {
    return combineLatest([this.dbGetAll$(), this.selectedMonth$]).pipe(
      filter(([expenses, selectedMonth]) => !!(expenses && selectedMonth)),
      map(([expenses, selectedMonth]) => {
        return expenses.filter((e) => moment(e.date).isSame(selectedMonth, 'month'));
      })
    );
  }

  getExpensesTotalAmount$(): Observable<number> {
    return this.dbGetAllFilteredBySelectedMonth$().pipe(
      map((expenses) => expenses.reduce((acc, e) => acc + e.amount, 0))
    );
  }
}
