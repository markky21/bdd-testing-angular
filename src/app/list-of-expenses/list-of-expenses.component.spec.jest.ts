import { ComponentFixture, TestBed } from '@angular/core/testing';

import { ListOfExpensesComponent } from './list-of-expenses.component';
import { NO_ERRORS_SCHEMA } from '@angular/core';

// NOTE: If a component is only used to display an assembly of components,
// then we should do a test to see if these components are visible in the DOM.
// However, we don't need to bootstrap these components,
// so we should omit including dependencies in ModuleDef and set NO_ERRORS_SCHEMA

describe('ListOfExpensesComponent', () => {
  let fixture: ComponentFixture<ListOfExpensesComponent>;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [ListOfExpensesComponent],
      schemas: [NO_ERRORS_SCHEMA],
    }).compileComponents();

    fixture = TestBed.createComponent(ListOfExpensesComponent);
    fixture.detectChanges();
  });

  it('should display page components', () => {
    expect(fixture).toMatchSnapshot();
  });
});
