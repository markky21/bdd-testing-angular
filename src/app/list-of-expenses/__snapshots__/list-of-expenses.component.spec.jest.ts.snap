// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListOfExpensesComponent should display page components 1`] = `
<app-list-of-expenses>
  <mat-card
    class="Wrapper"
  >
    <mat-card-header>
      <mat-card-title>
        List of expenses
      </mat-card-title>
    </mat-card-header>
    <mat-card-content>
      <app-month-selector />
      <app-expense-table />
      <app-expense-form />
    </mat-card-content>
  </mat-card>
</app-list-of-expenses>
`;
